### ISSUE TITLE
Option to pin Asciidoctor PDF version

### ISSUE THREAD

**Issue by infotexture** [asciidoctor/asciidoctor-intellij-plugin#1155]
## Why the new feature should be added

Current versions of the plug-in automatically download Asciidoctor PDF version 2.3 when the **PDF** button is clicked in the toolbar, but due to recent changes in the Asciidoctor PDF theme format, themes created for earlier Asciidoctor PDF versions throw ~~errors~~ _warnings_ with the latest version.

## How the new feature should work

If the plug-in could be configured to use a different version of Asciidoctor PDF, it would be possible to build PDF output for projects that have not yet migrated their themes to the latest Asciidoctor PDF versions.

This might also be a solution for #1095.

## Possible alternatives

Downgrade the locally installed version of the plug-in to a [previous version](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/main/CHANGELOG.adoc) that bundles the desired version of Asciidoctor PDF.

**Comment by mojavelinux** [comment]
> throw errors with the latest version.

I don't think it's accurate to say that Asciidoctor PDF is throwing errors. It is issuing warnings. Asciidoctor PDF was carefully changed so as to support themes that worked as of the first stable release (1.5.0). I'm not opposed to the version pinning idea as we will continue to release new major versions, and there will be a time when deprecated features get removed.

**Comment by infotexture** [comment]
@mojavelinux Thanks for chiming in. 🙏 
Didn't mean to suggest Asciidoctor PDF was doing anything wrong, apologies if it sounded like that. 🙇 

We're just feeding it a theme that was designed for 1.6.2, and it won't build with 2.3. We know we have to work our way through the changelogs to migrate our theme to the syntax that the new version expects, and we plan to do so soon. 

Until then, we're able to work around the issue by building with Asciidoctor PDF 1.6.2 on the CLI, or downgrading the installed IntelliJ AsciiDoc Plugin to the last version that bundled Asciidoctor PDF 1.6.2.

Just filed this issue thinking we may not be the only ones that need to build PDF output with older themes that haven't yet been migrated to the latest greatest syntax.

If you think themes that worked with 1.6.2 _should_ work with 2.3, I'd be happy to share ours, but that's a discussion we should probably have elsewhere, as it's beyond the scope of this issue.

**Comment by ahus1** [comment]
Hi @infotexture, thanks for sharing this. With the recent speed up of new versions I can see that people want to use an the version they use in their build also in their IDE when clicking in "PDF". And I agree the error handling in the IDE is missing at the moment, as stated in the issue you linked. Instead of showing an error message it creates HTML file with the error message.

I'll have a look on how to make this configurable, still am an not sure how and how far to take this. I see you named the version of AsciidoctorPDF you would like to use, and I see that this would mean to support a wide range or versions. I would also need to come up with some testing, let's see.

Something I try to avoid is manual configuration on the user's side. The plugin should determine the right thing to do by looking at the repository. In the setup you describe, how could the plugin do that?

**Comment by mojavelinux** [comment]
> If you think themes that worked with 1.6.2 should work with 2.3, I'd be happy to share ours, but that's a discussion we should probably have elsewhere, as it's beyond the scope of this issue.

If Asciidoctor PDF is failing, please report that. If the theme works with 1.6.2, then it should work with 2.3. I'll investigate and make a change if necessary.

**Comment by infotexture** [comment]
@ahus1 wrote:
> Something I try to avoid is manual configuration on the user's side. 
> The plugin should determine the right thing to do by looking at the repository.
> In the setup you describe, how could the plugin do that?

Makes sense. In our setup, there is currently no way to determine the Asciidoctor PDF version from the repo contents, but I suppose we could add a custom attribute to our `.asciidoctorconfig.adoc` or something like that if there's a best practice for this sort of metadata storage.

**Comment by infotexture** [comment]
@mojavelinux wrote:
> If Asciidoctor PDF is failing, please report it to the Asciidoctor PDF issue tracker.

Will do. For the record here, our custom theme works fine when built from the Asciidoctor IntelliJ plugin plug-in v0.36.13 (which bundles Asciidoctor 1.6), but fails with a fatal error w/ plugin version 0.37.47 (`asciidoctorj-pdf v2.3.0`):

    Message during rendering index.adoc
    FATAL null:null: org.jruby.exceptions.NoMethodError: 
    (NoMethodError) undefined method `-' for #  Did you mean? -@

I'll follow up by filing an Asciidoctor PDF issue, and include our theme file there for debugging.

**Comment by ahus1** [comment]
>  there is currently no way to determine the Asciidoctor PDF version from the repo contents

Hm, I was hoping to read it from a Gradle or Maven file. I don't want to attributes for that as that are is already quite busy with pre-defined attributes. Let me think what's possible, it might take a while. Maybe the plugin will finally get project specific configurations in the IntelliJ way that can (if you wish to) be checked in to the `.idea` folder.

**Comment by infotexture** [comment]
@ahus1 Thanks for your consideration, but don't put too much effort into this if it's not something commonly requested.

We've traced the cause of our issues with Asciidoctor PDF 2.3.0 in https://github.com/asciidoctor/asciidoctor-pdf/issues/2343. The problem was solved with a one-line fix to our existing theme, and the next 2.3.x release will include a backwards-compatibility fix that resolves the issue without requiring changes to older themes.

Feel free to close this issue, as we've found another solution.

Thanks for your work. 🙏
